### PR TITLE
azuredevops: Support allowOrgs in azure auth provider

### LINKED ISSUE #48555: Support `allowOrgs` in AzureDevOps auth provider
We want customers to be able to limit users from signing up only if they are a part of a pre-defined list of orgs. 

- [ ] Implement list user orgs on the azuredevops Client
- [ ] Use the client method during the OAuth signup / login step to check if user is part of the `allowOrgs` list before allowing the signup / login to succeed

### PR DESCRIPTION
**Pull request by indradhanush**
Fixes #48555.

## Test plan

- Added test
- Tested locally
- Builds should pass

### CODE REVIEW

File: enterprise/cmd/frontend/internal/auth/azureoauth/session.go [hunk 1]
```diff
@@ -31,6 +36,13 @@ func (s *sessionIssuerHelper) GetOrCreateUser(ctx context.Context, token *oauth2
 		return nil, "failed to read Azure DevOps Profile from oauth2 callback request", errors.Wrap(err, "azureoauth.GetOrCreateUser: failed to read user from context of callback request")
 	}
 
+	if allow, err := s.verifyAllowOrgs(ctx, token); err != nil {
+		return nil, "error in verifying authorized user organizations", err
+	} else if !allow {
+		msg := "user is not part of authorized orgs list"
```
Comment (kopancek) [inline enterprise/cmd/frontend/internal/auth/azureoauth/session.go:42]: We have a different [error message for Github](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/cmd/frontend/internal/auth/githuboauth/session.go?L73), let's make them sound similar:
```suggestion
		msg := "user does not belong to allowed organizations"
```

File: enterprise/cmd/frontend/internal/auth/azureoauth/session.go [hunk 2]
```diff
@@ -31,6 +36,13 @@ func (s *sessionIssuerHelper) GetOrCreateUser(ctx context.Context, token *oauth2
...
+		return nil, msg, errors.Newf("%s, must be in one of %v", msg, s.allowOrgs)
```
Comment (kopancek) [inline enterprise/cmd/frontend/internal/auth/azureoauth/session.go:43]: I think in the other providers, for security reasons we do not show which groups those are. Might be worth checking with security if this is OK.

File: enterprise/cmd/frontend/internal/auth/azureoauth/session.go [hunk 3]
```diff
@@ -97,6 +108,47 @@ func (s *sessionIssuerHelper) AuthFailedEventName() database.SecurityEventName {
...
+
+	authorizedOrgs, err := client.ListAuthorizedUserOrganizations(ctx, profile)
+	if err != nil {
+		return false, errors.Wrap(err, "failed to list organizations of user")
+	}
+
+	allowedOrgs := map[string]struct{}{}
+	for _, org := range s.allowOrgs {
+		allowedOrgs[org] = struct{}{}
+	}
```
Comment (kopancek) [inline enterprise/cmd/frontend/internal/auth/azureoauth/session.go:141]: Should we instead define `s.allowOrgs` as a set? This seems inefficient, to always do the conversion even if the config did not change.

File: enterprise/cmd/frontend/internal/auth/azureoauth/session.go [hunk 4]
```diff
@@ -97,6 +108,47 @@ func (s *sessionIssuerHelper) AuthFailedEventName() database.SecurityEventName {
...
+
+	for _, org := range authorizedOrgs {
+		if _, ok := allowedOrgs[org.Name]; ok {
+			return true, nil
+		}
+	}
```
Comment (kopancek) [inline enterprise/cmd/frontend/internal/auth/azureoauth/session.go:147]: Again, dreaming of `Intersection[T](a set[T], b set[T])` that would give you answer immediately :D

File: enterprise/cmd/frontend/internal/auth/azureoauth/session.go (hunk 2 above)
Comment (indradhanush) [reply enterprise/cmd/frontend/internal/auth/azureoauth/session.go:43]: Yes, we're not showing it here either, The `safeErrMsg` does not contain the list. Only the error which is not shown to the end user.

Comment (indradhanush) [reply enterprise/cmd/frontend/internal/auth/azureoauth/session.go:43]: Here's what the failure looks like to the user:

image

File: enterprise/cmd/frontend/internal/auth/azureoauth/session.go [hunk 5]
```diff
@@ -11,17 +11,22 @@ import (
...
+const (
+	stateCookie         = "azure-state-cookie"
+	urnAzureDevOpsOAuth = "AzureDevOpsOAuth"
+)
 
 type sessionIssuerHelper struct {
 	*extsvc.CodeHost
 	db          database.DB
 	clientID    string
+	allowOrgs   []string
```
Comment (pjlast) [inline enterprise/cmd/frontend/internal/auth/azureoauth/session.go]: Store these orgs as a `map[string]struct{}{}` to begin with and you won't have to do the double loop conversion later down on every request

File: enterprise/cmd/frontend/internal/auth/azureoauth/session.go (hunk 2 above)
Comment (indradhanush) [reply enterprise/cmd/frontend/internal/auth/azureoauth/session.go]: And now with the updated user message like this:

image

File: enterprise/cmd/frontend/internal/auth/azureoauth/session.go (hunk 3 above)
Comment (indradhanush) [reply enterprise/cmd/frontend/internal/auth/azureoauth/session.go]: Makes sense. Fixed.

File: enterprise/cmd/frontend/internal/auth/azureoauth/session_test.go [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+	"github.com/sourcegraph/sourcegraph/internal/extsvc/azuredevops"
+	"github.com/sourcegraph/sourcegraph/internal/rcache"
+	"golang.org/x/oauth2"
+)
+
+func Test_verifyAllowOrgs(t *testing.T) {
+	rcache.SetupForTest(t)
+
+	mockServerInvokedCount := 0
```
Comment (sashaostrikov) [inline enterprise/cmd/frontend/internal/auth/azureoauth/session_test.go:20]: oh no, is there no built-in way of counting invocations of mocked HTTP server? :(

**Comment by pjlast** [comment]
Welp by the time I submitted you applied most of the changes already 😆

File: enterprise/cmd/frontend/internal/auth/azureoauth/session.go (hunk 5 above)
Comment (indradhanush) [reply enterprise/cmd/frontend/internal/auth/azureoauth/session.go]: Just fixed. :D

File: enterprise/cmd/frontend/internal/auth/azureoauth/session_test.go [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+	for _, tc := range testCases {
+
+		t.Run(tc.name, func(t *testing.T) {
+			mockServerInvokedCount = 0
+			s := &sessionIssuerHelper{allowOrgs: tc.allowOrgs}
+
+			ctx := context.Background()
+			allow, err := s.verifyAllowOrgs(ctx, &oauth2.Token{AccessToken: "foo"})
+			if err != nil {
```
Comment (sashaostrikov) [inline enterprise/cmd/frontend/internal/auth/azureoauth/session_test.go:110]: such checks can be substituted with 1-liner `require` invocations, like:
`require.NoError(err)`

**Review by sashaostrikov — APPROVED** [review]
LGTM overall, left a couple of comments :)

File: enterprise/cmd/frontend/internal/auth/azureoauth/session.go [hunk 8]
```diff
@@ -97,6 +109,42 @@ func (s *sessionIssuerHelper) AuthFailedEventName() database.SecurityEventName {
 	return database.SecurityEventAzureDevOpsAuthFailed
 }
 
+func (s *sessionIssuerHelper) verifyAllowOrgs(ctx context.Context, token *oauth2.Token) (bool, error) {
+	if len(maps.Keys(s.allowOrgs)) == 0 {
+		return true, nil
+	}
```
Comment (pjlast) [inline enterprise/cmd/frontend/internal/auth/azureoauth/session.go:115]: You don't need to do `len(maps.Keys(s.allowOrgs))` you can just do `len(s.allowOrgs)`

File: enterprise/cmd/frontend/internal/auth/azureoauth/session.go [hunk 9]
```diff
@@ -31,6 +36,13 @@ func (s *sessionIssuerHelper) GetOrCreateUser(ctx context.Context, token *oauth2
 		return nil, "failed to read Azure DevOps Profile from oauth2 callback request", errors.Wrap(err, "azureoauth.GetOrCreateUser: failed to read user from context of callback request")
 	}
 
+	if allow, err := s.verifyAllowOrgs(ctx, token); err != nil {
```
[the remark this answers is not in the document]
Comment (indradhanush) [reply enterprise/cmd/frontend/internal/auth/azureoauth/session.go:39]: Good suggestion. Fixed.

File: enterprise/cmd/frontend/internal/auth/azureoauth/session.go [hunk 10]
```diff
@@ -97,6 +108,47 @@ func (s *sessionIssuerHelper) AuthFailedEventName() database.SecurityEventName {
...
+		nil,
+	)
+	if err != nil {
+		return false, errors.Wrap(err, "failed to create client for listing organizations of user")
+	}
+
+	profile, err := client.GetAuthorizedProfile(ctx)
+	if err != nil {
+		return false, errors.Wrap(err, "failed to get authorized user profile")
+	}
```
[the remark this answers is not in the document]
Comment (pjlast) [reply enterprise/cmd/frontend/internal/auth/azureoauth/session.go:131]: Just still curious about this one here :) other than that it looks good to me

**Review by pjlast — APPROVED** [review]
Just wondering if we can get rid of one last request but otherwise :shipit:

File: enterprise/cmd/frontend/internal/auth/azureoauth/session.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (indradhanush) [reply enterprise/cmd/frontend/internal/auth/azureoauth/session.go]: This is a great catch. I totally missed this - saved an API call! Thank you.

Fixed.

**Comment by indradhanush** [comment]
> Just wondering if we can get rid of one last request but otherwise :shipit:

Ask and ye shall receive.

**Comment by indradhanush** [comment]
Hmm. Looks like something broke with the latest commit to reuse the profile. Now it always fails. Looking into this.

**Comment by indradhanush** [comment]
Lol. I had a typo in my site config and misnamed the org in the `allowOrgs`. 🤦🏽‍♂️

All good now.

**Pull request merged**